### PR TITLE
Allow alternate implementation of GCM

### PR DESCRIPTION
**Pull request by Patater**
Provide the ability to use an alternative implementation of GCM in place
of the library provide implementation.

NOTES:
- This is a new feature and does not need to be backported.
- This does not follow Ron's `ALT` style (https://github.com/ARMmbed/mbedtls/pull/867)
- Tested with "alternate" software implementation of GCM that's actually just the library implementation (https://github.com/Patater/mbedtls/commits/gcm-alt-test)

```
test_suite_aes.cbc ................................................ PASS
test_suite_aes.cfb ................................................ PASS
test_suite_aes.ecb ................................................ PASS
test_suite_aes.rest ............................................... PASS
test_suite_arc4 ................................................... PASS
test_suite_asn1write .............................................. PASS
test_suite_base64 ................................................. PASS
test_suite_blowfish ............................................... PASS
test_suite_camellia ............................................... PASS
test_suite_ccm .................................................... PASS
test_suite_cipher.aes ............................................. PASS
test_suite_cipher.arc4 ............................................ PASS
test_suite_cipher.blowfish ........................................ PASS
test_suite_cipher.camellia ........................................ PASS
test_suite_cipher.ccm ............................................. PASS
test_suite_cipher.des ............................................. PASS
test_suite_cipher.gcm ............................................. PASS
test_suite_cipher.null ............................................ PASS
test_suite_cipher.padding ......................................... PASS
test_suite_cmac ................................................... PASS
test_suite_ctr_drbg ............................................... PASS
test_suite_debug .................................................. PASS
test_suite_des .................................................... PASS
test_suite_dhm .................................................... PASS
test_suite_ecdh ................................................... PASS
test_suite_ecdsa .................................................. PASS
test_suite_ecjpake ................................................ PASS
test_suite_ecp .................................................... PASS
test_suite_entropy ................................................ PASS
test_suite_error .................................................. PASS
test_suite_gcm.aes128_de .......................................... PASS
test_suite_gcm.aes128_en .......................................... PASS
test_suite_gcm.aes192_de .......................................... PASS
test_suite_gcm.aes192_en .......................................... PASS
test_suite_gcm.aes256_de .......................................... PASS
test_suite_gcm.aes256_en .......................................... PASS
test_suite_gcm.camellia ........................................... PASS
test_suite_hmac_drbg.misc ......................................... PASS
test_suite_hmac_drbg.no_reseed .................................... PASS
test_suite_hmac_drbg.nopr ......................................... PASS
test_suite_hmac_drbg.pr ........................................... PASS
test_suite_md ..................................................... PASS
test_suite_mdx .................................................... PASS
test_suite_memory_buffer_alloc .................................... PASS
test_suite_mpi .................................................... PASS
test_suite_pem .................................................... PASS
test_suite_pk ..................................................... PASS
test_suite_pkcs1_v15 .............................................. PASS
test_suite_pkcs1_v21 .............................................. PASS
test_suite_pkcs5 .................................................. PASS
test_suite_pkparse ................................................ PASS
test_suite_pkwrite ................................................ PASS
test_suite_rsa .................................................... PASS
test_suite_shax ................................................... PASS
test_suite_ssl .................................................... PASS
test_suite_timing ................................................. PASS
test_suite_version ................................................ PASS
test_suite_x509parse .............................................. PASS
test_suite_x509write .............................................. PASS
test_suite_xtea ................................................... PASS
------------------------------------------------------------------------
PASSED (60 suites, 6109 tests run)
```

### CODE REVIEW

**Comment by andresag01** [comment]
@Patater: Could you please add a ChangeLog entry for this new feature? Thanks!

**Comment by Patater** [comment]
@andresag01 Done

File: include/mbedtls/gcm.h [hunk 1]
```diff
@@ -217,4 +219,8 @@ int mbedtls_gcm_self_test( int verbose );
 }
 #endif
 
+#else  /* !MBEDTLS_GCM_ALT */
+#include "gcm_alt.h"
+#endif /* !MBEDTLS_GCM_ALT */
```
Comment (andresag01) [inline include/mbedtls/gcm.h]: I do not think that we should include the function `int mbedtls_gcm_self_test( int verbose )` inside `MBEDTLS_GCM_ALT` as this is simply a test and porters will be just copying it as it is in the .c file.

Comment (Patater) [reply include/mbedtls/gcm.h]: Yup, good catch. I made sure to include it in `gcm.c` (whether or not `MBEDTLS_GCM_ALT` is used), but didn't declare the prototype in `gcm.h` if `MBEDTLS_GCM_ALT` was defined.

**Comment by RonEld** [comment]
HI @Patater Why doesn't this follow my `ALT` style as in #867?
This new style is for the following reasons:
1. If alternative implementation doesn't support a specfic feature (such as an unsupported keysize), we should allow usage of SW fallback, without copying the code to the alternative implementor's code.
2. To avoid copying the API to the `xxx_alt.h` which is redundant, and might cause future issues, if API happens to change ( as in https://github.com/ARMmbed/mbedtls-restricted/pull/357)

I strongly recommend changing the code to the same style, which was already internally discussed and agreed, and this will save us future work

File: include/mbedtls/gcm.h [hunk 2]
```diff
@@ -217,4 +231,5 @@ int mbedtls_gcm_self_test( int verbose );
 }
 #endif
 
+
```
Comment (yanesca) [inline include/mbedtls/gcm.h]: Why do we need this extra line?

Comment (Patater) [reply include/mbedtls/gcm.h]: I would guess we don't need it. :)

**Comment by gilles-peskine-arm** [comment]
CI failed only due to known unrelated issues (compat.sh slow start and timing tests). Ok to merge.

**Pull request merged**